wkcuber/api/View.py

```diff
@@ -158,6 +158,79 @@ def for_each_chunk(
         # execute the work for each chunk
         wait_and_ensure_success(executor.map_to_futures(work_on_chunk, job_args))
 
+    def for_zipped_chunks(
+        self,
+        work_on_chunk: Callable[[List[Any]], None],
+        job_args_per_chunk: Any,
+        target_view: "View",
+        source_chunk_size: Tuple[int, int, int],
+        target_chunk_size: Tuple[int, int, int],
+        executor: Union[ClusterExecutor, cluster_tools.WrappedProcessPoolExecutor],
+    ) -> None:
+        """
+        This method is similar to 'for_each_chunk' in the sense, that it delegates work to smaller chunks.
+        However, this method also takes another view as a parameter. Both views are chunked simultaneously
+        and a matching pair of chunks is then passed to the function that shall be executed.
+        This is useful if data from one view should be (transformed and) written to a different view,
+        assuming that the transformation of the data can be handled on chunk-level.
+        Additionally to the two views and the job_args, the counter 'i' is passed to the 'work_on_chunk',
+        which can be used for logging.
+        """
+        source_offset = np.array(self.global_offset)
+        target_offset = np.array(target_view.global_offset)
+        source_chunk_size_np = np.array(source_chunk_size)
+        target_chunk_size_np = np.array(target_chunk_size)
+        assert not np.any(
+            source_chunk_size_np % target_chunk_size_np
+        ), f"Calling 'for_zipped_chunks' failed because the source_chunk_size ({source_chunk_size} must be divisible by the target_chunk_size ({target_chunk_size}))"
+        source_to_target_scale_np = source_chunk_size_np // target_chunk_size_np
+
+        calculated_target_size = -(
+            -np.array(self.size) // source_to_target_scale_np
+        )  # ceil div
+
+        if not target_view.check_bounds(target_offset, calculated_target_size):
+            raise AssertionError(
+                f"Calling 'for_zipped_chunks' failed because the size of the target_view ({target_view.size}) must be at least {calculated_target_size} or the view must be not bounded."
+            )
+
+        job_args = []
+        source_chunks = BoundingBox(source_offset, self.size).chunk(
+            source_chunk_size_np, list(source_chunk_size_np)
+        )
+        target_chunks = BoundingBox(target_offset, calculated_target_size).chunk(
+            target_chunk_size, list(target_chunk_size)
+        )
+
+        for i, (source_chunk, target_chunk) in enumerate(
+            zip(source_chunks, target_chunks)
+        ):
```

Comment:
I agree with you that the current implementation is not optimal. 
I think the mapping of chunks from one view to the other should be bijective (no source/target chunks should appear twice). This allows two use cases:
1. Both, the `target_view` and the `source_view`, have the same size and their `chunk_sizes` also match (e.g. copying data from A to B). (This is a special case of 2.)
2. The `source_view` and `target_view` can have different sizes (e.g. downsampling: the `source_view` is larger than `target_view` by a factor of `scale`). In this case, the ratio of the `chunk_size` of the views must be the same as the ratio between the sizes of the views.

This ensures that number of chunks for both views is the same. Therefore, `zip` works as expected.
I would propose, that I update the assertions in the function to the behavior described above.

Your given example is in my opinion not valid because `small_chunk_1` and `small_chunk_2` could be combined to one chunk, by choosing a different `target_chunk_size`. Is there a use case where your example would be nice and the behavior of above does not work?

Another thing that concerns me a little with this method is that the user has to be a bit careful with choosing the chunk_sizes: Let's say that data should be written to the `target_view`. The work on the different chunks is executed in parallel. This means that the `target_chunk_size` must be a aligned with the files on disk to avoid concurrent writes. Currently there is no such assertion because the `View` has no information about how the data is stored on disk.

Reply:
Cool, that clears some things up for me :+1: Somehow I was stuck with the concept that the chunk size would be the cube size in which the wkw file is. With that assumption, it was hard to wrap my head around how the mapping would be bijective. However, your explanation makes total sense and I also think that this abstraction is more elegant to use.

Could you add that bit to the docstring of the method? Mainly, that the mapping is bijective and how this can be used to do stuff, such as downsampling.

> I would propose, that I update the assertions in the function to the behavior described above.

Yes, that would be good.

> Currently there is no such assertion because the View has no information about how the data is stored on disk.

Hm, I think, that would be a good thing to have. Otherwise, we will run into this sooner or later and have a hard time debugging this. Can we allow the `View` to access the header information somehow?

Reply:
I was wrong, the `View` already had the header. I have added the assertions only for the `target_view` because in my vision data should only be written to the `target_view` (as the name suggests). However, currently nothing prevents the user from also writing things to the `source_view`. This could be solved by introducing an optional parameter `read_only` to the function `get_view`. Do you think this is necessary?

Reply:
I think it would be better to be on the safe side. A `read_only` flag sounds like a very nice solution for this. If it's easy to implement, I'd vote for that :)